server/handlers/validation.go

```diff
@@ -356,158 +373,90 @@ func validateTargets(role string, roles map[string]storage.MetaUpdate, repo *tuf
-		//              for rotations? Should the user have to provide a flag
-		//              which gets transmitted to force a root update without
-		//              correct old key signatures.
+	newTimestampRole, err := parsedNewRoot.BuildBaseRole(data.CanonicalTimestampRole)
+	if err != nil { // should never happen, since the root metadata has been validated
 		return nil, err
 	}
+	// According to the TUF spec, any role may have more than one signing
+	// key and require a threshold signature.  However, notary-server
+	// creates the timestamp, and there is only ever one, so a threshold
+	// greater than one would just always fail validation
+	if newTimestampRole.Threshold != 1 {
+		return nil, fmt.Errorf("timestamp role has invalid threshold")
+	}
 
-	if !data.ValidTUFType(parsedNewRoot.Signed.Type, data.CanonicalRootRole) {
-		return nil, fmt.Errorf("root has wrong type")
+	if oldRoot != nil {
+		if err := checkAgainstOldRoot(oldRoot, newRootRole, parsedNewSigned); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := signed.VerifyRoot(parsedNewSigned, newRootRole.Threshold, newRootRole.Keys); err != nil {
+		return nil, err
 	}
+
 	return parsedNewRoot, nil
 }
 
-// checkRoot errors if an invalid rotation has taken place, if the
-// threshold number of signatures is invalid, if there are an invalid
-// number of roles and keys, or if the timestamp keys are invalid
-func checkRoot(oldRoot, newRoot *data.SignedRoot, timestampKey data.PublicKey) error {
-	rootRole := data.CanonicalRootRole
-	targetsRole := data.CanonicalTargetsRole
-	snapshotRole := data.CanonicalSnapshotRole
-	timestampRole := data.CanonicalTimestampRole
-
-	var oldRootRole *data.RootRole
-	newRootRole, ok := newRoot.Signed.Roles[rootRole]
-	if !ok {
-		return errors.New("new root is missing role entry for root role")
+// checkAgainstOldRoot errors if an invalid root rotation has taken place
+func checkAgainstOldRoot(oldRoot []byte, newRootRole data.BaseRole, newSigned *data.Signed) error {
+	parsedOldRoot := &data.SignedRoot{}
+	err := json.Unmarshal(oldRoot, parsedOldRoot)
+	if err != nil {
+		logrus.Warn("Old root could not be parsed, and cannot be used to check the new root.")
+		return nil
 	}
 
-	oldThreshold := 1
-	rotation := false
-	oldKeys := map[string]data.PublicKey{}
-	newKeys := map[string]data.PublicKey{}
-	if oldRoot != nil {
-		// check for matching root key IDs
-		oldRootRole = oldRoot.Signed.Roles[rootRole]
-		oldThreshold = oldRootRole.Threshold
-
-		for _, kid := range oldRootRole.KeyIDs {
-			k, ok := oldRoot.Signed.Keys[kid]
-			if !ok {
-				// if the key itself wasn't contained in the root
-				// we're skipping it because it could never have
-				// been used to validate this root.
-				continue
-			}
-			oldKeys[kid] = data.NewPublicKey(k.Algorithm(), k.Public())
-		}
-
-		// super simple check for possible rotation
-		rotation = len(oldKeys) != len(newRootRole.KeyIDs)
+	oldRootRole, err := parsedOldRoot.BuildBaseRole(data.CanonicalRootRole)
```

Comment:
Note that this is a slight change in behavior:  previously, if the root had key IDs that didn't belong to any key, we would have counted that as a valid old root and used it to validate the new root.  Now, because we consider that an invalid root role, we will not use the old root to validate the new root.  Should I just build the base role without checking for that?